### PR TITLE
[SQL][SPARK-6785] fix DateUtils.fromJavaDate(java.util.Date) for Dates before 1970

### PR DESCRIPTION
**Pull request by ckadner**
BUG: With the current implementation, the from-to-Java date conversion will be off by one day for Dates before 1970 because of a rounding (truncate) flaw in the function DateUtils#millisToDays(Long).

FIX: The fix is to do the conversion using Double and flooring the fractions instead of truncating them.

NOTE: Before this fix, the code DID work for Dates that are NOT one millisecond before or after midnight in the system's local time zone.

### CODE REVIEW

**Comment by davies** [comment]
OK to test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateUtils.scala [hunk 1]
```diff
@@ -41,8 +41,9 @@ object DateUtils {
   }
 
   // we should use the exact day as Int, for example, (year, month, day) -> day
-  def millisToDays(millisLocal: Long): Int = {
-    ((millisLocal + LOCAL_TIMEZONE.get().getOffset(millisLocal)) / MILLIS_PER_DAY).toInt
+  def millisToDays(millisUtc: Long): Int = {
+    // SPARK-6785: use Math.floor so negative number of days (dates before 1970) work as input for function toJavaDate(Int)
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateUtils.scala]: This line is too long

**Comment by davies** [comment]
Could you add `[SQL]` in the title? also follow the rules for title as others.

**Comment by davies** [comment]
@ckadner Thanks for working on this, the patch looks good to me, just one minor comment.

**Comment by ckadner** [comment]
Thanks for your review Davies! I corrected the title and split the offending long line of code in 2.

**Comment by davies** [comment]
LGTM, waiting for tests.

**Comment by ckadner** [comment]
Apologies for missing the explicit method return type. I just corrected that.

**Comment by davies** [comment]
The unit test is outdated, please update it: ScalaReflectionRelationSuite: query case class RDD

**Comment by ckadner** [comment]
Thank you Davies for hanging in there with me. And apologies that I did not catch this one myself. I just fixed the unit test and the remaining unit tests in Catalyst, Hive, SQL are running through for me.

**Comment by rxin** [comment]
Jenkins, test this please.

**Comment by ckadner** [comment]
Hi Reynold (@rxin)
can you help me make sense of the last build log? **_...fails MiMa tests_**  Does this look like a configuration error? I do not readily see a connection to my code changes or how I could run this test suite on my own to root out possible problems.
Thank you in advance!

**Comment by srowen** [comment]
@ckadner A MiMa failure means it thinks something about the change modified a public API. It can be subtle sometimes. Here I can't pick out the actual MiMa error though. Try `./dev/mima` locally to see if it complains about anything.

**Comment by davies** [comment]
@ckadner  There is one more test should be updated: Date comparison test 2

**Comment by ckadner** [comment]
@srowen ...so far I have no luck with running ./dev/mima 

@davies ...I propose to remove the test case _"Date comparison test 2"_ since the expression `"CAST(CAST(0 AS timestamp) AS date) > CAST(0 AS timestamp)"` will produce different results based on the system timezone of the machine running it, 

i.e. it will evaluate to `true` in time zones east of GMT, i.e. Berlin, Germany ...

```
 "CAST(0 AS timestamp)" --> "1 Jan 1970 00:00:00 GMT"
 "CAST(CAST(0 AS timestamp) AS date)" --> "1 Jan 1970 01:00:00 GMT"
```
... but to `false` west of GMT, i.e. San Francisco ...
```
 "CAST(0 AS timestamp)" --> "1 Jan 1970 00:00:00 GMT"
 "CAST(CAST(0 AS timestamp) AS date)" --> "31 Dec 1969 08:00:00 GMT"
```
Note, this conversion is only accurate down to the _day_ of the `Date` object, since DateUtils uses full days to convert dates and the back-and-forth timezone adjustment

**Comment by ckadner** [comment]
@srowen, @davies, thank you for your pointers!  I will work further on to this tomorrow.

**Comment by ckadner** [comment]
Sorry, it took me some time to get back.

Turns out Hive suffers from the same bug ([HIVE-10178](https://issues.apache.org/jira/browse/HIVE-10178)) which causes the test cases based on `HiveComparisonTest` to fail. The Hive bug was fixed in _Hive 1.2.0._ and the code (and fix) in Hive's `DateWritable.java` looks remarkably similar to Spark's `DateUtils`.

@srowen, @davies, I need your guidance:
- how (and how often) do fixes in Hive get consumed by Spark? 
- should we move on with this fix in Spark and tweak the `HiveComparisonTest` to tolerate the discrepancy until Spark consumes Hive 1.2.0 (or another branch/version) that has the [HIVE-10178](https://issues.apache.org/jira/browse/HIVE-10178) fix?

Thanks in advance!

**Comment by srowen** [comment]
Right now Spark is on Hive 0.12 / 0.13 although it will become more possible to use later Hive in the next version. So, that means our Hive comparison tests are going to get more fragile I think, if Hive behavior is changing. Is there any way to write this to expect several possible correct behaviors?

**Comment by ckadner** [comment]
@srowen, thanks for your advice!

I modified the `HiveQuerySuite` to execute different test cases based on the version of Hive that is used internally by Spark SQL. It's not very elegant, but this way the Date cast tests will work now and after Spark upgrades to use Hive 1.2.0 (or later).

**Comment by marmbrus** [comment]
We will only ever run the compatibility tests with the execution version of Hive.  So, I don't think it'll get any more brittle, but we will need to update the list/tests when we upgrade (WIP PR here #6748).

In any case where hive is returning the wrong answer, we usually just add the test case to the blacklist (in HiveCompatibilitySuite).  Since this is just our own test case tough, I'd move it to SQLQuerySuite and use `checkAnswer` instead.

**Comment by marmbrus** [comment]
ping, do you think you could bring this up to date?

**Comment by ckadner** [comment]
Hi Michael, 
for the failing hive comparison tests `"Date cast"` & `"Date comparison test 2"` I will add equivalent test cases to `SQLQuerySuite` probably in a new PR. I take it you want me to remove these 2 broken tests from `HiveQuerySuite` (as opposed to commenting them out) and delete the corresponding golden answer files?
Do we ever bring test cases like these back? Should I open a new JIRA to reinstate these test cases once [SPARK-8064](https://issues.apache.org/jira/browse/SPARK-8064) is resolved and the execution version of Hive is 1.2 ?

**Comment by marmbrus** [comment]
Yeah you can go ahead and remove then and the golden files.  I think your new tests will be sufficient and we wan't need to bring them back (HiveQueryTests are easier to write, but aren't inherently better than tests where you manually hard code the answers).

**Comment by ckadner** [comment]
Thanks Michael, I will open a new PR with a cleaned up version of my fix.

**Comment by ckadner** [comment]
I opened a new pull-request [PR-6983](https://github.com/apache/spark/pull/6983)

**Pull request closed without merging**